netket/_driver.py

```diff
@@ -0,0 +1,265 @@
+            n_samples: Number of Markov Chain Monte Carlo sweeps to be
+                performed at each step of the optimization.
+            discarded_samples (int): Number of sweeps to be discarded at the
+                beginning of the sampling, at each step of the optimization.
+                Defaults to 10% of n_samples.
+            discarded_samples_on_init (int): Number of sweeps to be discarded in
+                the first step of optimization, at the beginning of the
+                sampling.
+            sr (SR, optional): Determines whether and how stochastic reconfiguration
+                is applied to the bare energy gradient before performing applying
+                the optimizer. If this parameter is not passed or None, SR is not used.
+
+        Example:
+            Optimizing a 1D wavefunction with Variational Mante Carlo.
+
+            ```python
+            >>> import netket as nk
+            >>> SEED = 3141592
+            >>> g = nk.graph.Hypercube(length=8, n_dim=1)
+            >>> hi = nk.hilbert.Spin(s=0.5, graph=g)
+            >>> ma = nk.machine.RbmSpin(hilbert=hi, alpha=1)
+            >>> ma.init_random_parameters(seed=SEED, sigma=0.01)
+            >>> ha = nk.operator.Ising(hi, h=1.0)
+            >>> sa = nk.sampler.MetropolisLocal(machine=ma)
+            >>> sa.seed(SEED)
+            >>> op = nk.optimizer.Sgd(learning_rate=0.1)
+            >>> vmc = nk.variational.Vmc(hamiltonian=ha, machine=ma, sampler=sa,
+            ... optimizer=op, n_samples=500)
+            >>> print(ma.n_visible)
+            8
+            ```
+        """
+        self._ham = hamiltonian
+        self._machine = machine
+        self._sampler = sampler
+        self._sr = sr
+
+        self._optimizer_step, self._optimizer_desc = make_optimizer_fn(optimizer)
+
+        self._npar = machine.n_par
+        self._mc_data = None
+
+        self.n_samples = n_samples
+        self.n_discard = n_discard if n_discard else n_samples // 10
+
+        self._obs = {}
+
+        self.step_count = 0
+
+    def _get_mc_stats(self, op):
+        loc = _local_values(
+            self._mc_data.samples, self._mc_data.log_values, self._machine, op
+        )
+        return loc, _statistics(loc)
+
+    def advance(self, n_steps=1):
+        """
+        Performs a number of VMC optimization steps.
+
+        Args:
+            n_steps (int): Number of steps to perform.
+        """
+
+        def update_samples():
+            self._mc_data = _vmc.compute_samples(
+                self._sampler, self.n_samples, self.n_discard, der_logs="centered"
+            )
+
+        if not self._mc_data:
+            update_samples()
+
+        for _ in range(n_steps):
+            # Estimate energy
+            eloc, self._stats = self._get_mc_stats(self._ham)
+
+            # Estimate energy gradient
+            grad = _vmc.gradient_of_expectation(eloc, self._mc_data.der_logs)
```

Comment:
I think the easiest is just to combine (if possible) these two functions in something that computes both expectations and gradients of expectations 

```python 
eloc, grad = _vmc.gradient_of_expectation(self._ham, self._mc_data) 
```